pkg/snapshot/restorer/restorer.go

```diff
@@ -722,3 +769,87 @@ func verifySnapshotRevision(clientKV client.KVCloser, snap *brtypes.Snapshot) er
 	}
 	return nil
 }
+
+// getDecompressedSnapshotReadCloser passes the given ReadCloser through the
+// snapshot decompressor if the snapshot is compressed using a compression policy.
+// Also returns whether the snapshot was initially compressed or not, as well as
+// the compression policy used for compressing the snapshot.
+func getDecompressedSnapshotReadCloser(rc io.ReadCloser, snap *brtypes.Snapshot) (io.ReadCloser, bool, string, error) {
+	isCompressed, compressionPolicy, err := compressor.IsSnapshotCompressed(snap.CompressionSuffix)
+	if err != nil {
+		return nil, false, compressionPolicy, err
+	}
+
+	if isCompressed {
+		// decompress the snapshot
+		rc, err = compressor.DecompressSnapshot(rc, compressionPolicy)
+		if err != nil {
+			return nil, true, compressionPolicy, fmt.Errorf("unable to decompress the snapshot: %v", err)
+		}
+	}
+
+	return rc, true, compressionPolicy, nil
+}
+
+func (r *Restorer) readSnapshotContentsFromReadCloser(rc io.ReadCloser, snap *brtypes.Snapshot) ([]byte, error) {
+	startTime := time.Now()
+
+	rc, wasCompressed, compressionPolicy, err := getDecompressedSnapshotReadCloser(rc, snap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decompress delta snapshot %s : %v", snap.SnapName, err)
+	}
+
+	buf := new(bytes.Buffer)
+	bufSize, err := buf.ReadFrom(rc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse contents from delta snapshot %s : %v", snap.SnapName, err)
+	}
+
+	totalTime := time.Now().Sub(startTime).Seconds()
+	if wasCompressed {
+		r.logger.Infof("successfully fetched data of delta snapshot in %v seconds [CompressionPolicy:%v]", totalTime, compressionPolicy)
+	} else {
+		r.logger.Infof("successfully fetched data of delta snapshot in %v seconds", totalTime)
+	}
+
+	if bufSize <= sha256.Size {
+		return nil, fmt.Errorf("delta snapshot is missing hash")
+	}
+
+	sha := buf.Bytes()
+	data := sha[:bufSize-sha256.Size]
+	snapHash := sha[bufSize-sha256.Size:]
+
+	// check for match
+	h := sha256.New()
+	if _, err := h.Write(data); err != nil {
+		return nil, fmt.Errorf("unable to check integrity of snapshot %s: %v", snap.SnapName, err)
+	}
+
+	computedSha := h.Sum(nil)
+	if !reflect.DeepEqual(snapHash, computedSha) {
+		return nil, fmt.Errorf("expected sha256 %v, got %v", snapHash, computedSha)
+	}
+
+	return data, nil
+}
+
+func (r *Restorer) readSnapshotContentsFromFile(filePath string, snap *brtypes.Snapshot) ([]byte, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file %s for delta snapshot %s : %v", filePath, snap.SnapName, err)
+	}
+
+	return r.readSnapshotContentsFromReadCloser(file, snap)
+}
+
+// errorArrayToError takes an array of errors and returns a single concatenated error
+func errorArrayToError(errs []error) error {
```

Comment:
I don't think that's true. I've written unit tests now to confirm the correct working of the function, and also handled the case where `errs` is an empty array.

In the function, each error is appended to `errString`, which is then returned as an error type. Please confirm

Reply:
Maybe you can just remove this entire method and just use 
```go
errors.Join(errs...) // this returns an error
```

Reply:
We need golang 1.20 for that right? Since etcdbr still uses go 1.19, and I didn't want to move to go 1.20 since it might introduce some incompatible changes which then need to be handled, and I also didn't want to vendor a new library dependency to achieve this, I stuck to using this `errorArrayToError` way itself.

Reply:
Ok i assumed it has already moved. My bad. Ok we keep this for now and later remove this function when we move to go 1.20